### PR TITLE
Generator callout pass

### PR DESCRIPTION
**Pull request by prithayan**
-  This pass calls an external program for all the `rtl.module.generated` nodes.
-  The corresponding `rtl.generator.schema` node is used for the port names.
-  The value for each port name is parsed from the attribute on the `rtl.module.generated` nodes.
-  The `rtl.module.generated` node is removed from the IR.
-   The module is then declared as an external module, whose definition is present in a file with the module name.
-  The file name in which the module definition exists is added as an attribute to the external module declaration.

```
rtl.module.generated @FIRRTLMem_2_0_0_16_10_0_1_0, @FIRRTLMem(%rd_clock_0: i1, %rd_en_0: i1, %rd_addr_0: i4,
  %rd_clock_1: i1, %rd_en_1: i1, %rd_addr_1: i4) -> (%rd_data_0: i16, %rd_data_1: i16) attributes {depth = 10 : i64,
  numReadPorts = 2 : i32, numWritePorts = 0 : ui32, readLatency = 0 : ui32, readUnderWrite = 0 : ui32, width = 16 : ui32, writeLatency = 1 : ui32, numReadWritePorts = 1 : ui32}
```
Is lowered to::
```
 rtl.module.extern @FIRRTLMem_2_0_0_16_10_0_1_0
(%rd_clock_0: i1, %rd_en_0: i1, %rd_addr_0: i4, %rd_clock_1: i1, %rd_en_1: i1, %rd_addr_1: i4) -> (%rd_data_0: i16, %rd_data_1: i16) 
attributes {Filename = "FIRRTLMem_2_0_0_16_10_0_1_0.v"}
```

### CODE REVIEW

**Comment by GeorgeLyon** [comment]
This is a bit outside me wheelhouse, but it seems like this will run "find program by name" on every operation, which may be a bit expensive. Would it be possible to parameterize this pass with a generated identifier and a function?  Something like ```
```
auto memoryGenerator = findProgramByName(...);
createRTLGeneratorCalloutPass("FIRRTLMem", []{ run(memoryGenerator, <args>); })
```
This would also open up some interesting possibilities for bindings.

**Comment by teqdruid** [comment]
This feels like a task for the build system...

**Comment by prithayan** [comment]
> 
> 
> This is a bit outside me wheelhouse, but it seems like this will run "find program by name" on every operation, which may be a bit expensive. Would it be possible to parameterize this pass with a generated identifier and a function? Something like
> 
> ```
> auto memoryGenerator = findProgramByName(...);
> createRTLGeneratorCalloutPass("FIRRTLMem", []{ run(memoryGenerator, <args>); })
> ```
> 
> This would also open up some interesting possibilities for bindings.

@GeorgeLyon , this is called on every module op, and the generator program needs to be passed as a command line argument, so that we can use different generators. We can also filter based on the schema.

File: lib/Dialect/SV/Transforms/GeneratorCallout.cpp [hunk 1]
```diff
@@ -0,0 +1,159 @@
+//===- RTLLegalizeNames.cpp - RTL Name Legalization Pass ------------------===//
```
Comment (darthscsi) [inline lib/Dialect/SV/Transforms/GeneratorCallout.cpp:1]: Fix header.

File: lib/Dialect/SV/Transforms/GeneratorCallout.cpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+struct RTLGeneratorCalloutPass
+    : public sv::RTLGeneratorCalloutPassBase<RTLGeneratorCalloutPass> {
+  void runOnOperation() override;
+
+private:
+  DenseMap<StringRef, RTLGeneratorSchemaOp> generatorSchema;
```
Comment (darthscsi) [inline lib/Dialect/SV/Transforms/GeneratorCallout.cpp:38]: Is this used?  I don't think it is, nor should it be needed.

File: lib/Dialect/SV/Transforms/GeneratorCallout.cpp [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+};
+} // end anonymous namespace
+
+// Parse the \p genExec string, to extract the program executable, the path to
+// the executable and all the required arguments from \p genExecArgs.
+void static parseProgramArgs(const std::string genExec,
+                             const std::string genExecArgs,
+                             std::string &execName,
+                             std::string &genExecutablePath,
+                             std::vector<std::string> &genOptions) {
```
Comment (darthscsi) [inline lib/Dialect/SV/Transforms/GeneratorCallout.cpp:48]: Return a SmallVector by value

File: lib/Dialect/SV/Transforms/GeneratorCallout.cpp [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+  std::string temp = genExecArgs;
+  size_t pos;
+  while ((pos = temp.find_last_of(";")) != std::string::npos) {
```
Comment (darthscsi) [inline lib/Dialect/SV/Transforms/GeneratorCallout.cpp:51]: StringRef has a split by deliminter method, you should use it.

File: lib/Dialect/SV/Transforms/GeneratorCallout.cpp [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    genOptions.push_back(temp.substr(pos + 1));
+    temp = temp.substr(0, pos);
+  }
+  genOptions.push_back(temp);
+  temp = genExec;
```
Comment (darthscsi) [inline lib/Dialect/SV/Transforms/GeneratorCallout.cpp:56]: split out exec finding to a different function.

File: lib/Dialect/SV/Transforms/GeneratorCallout.cpp [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+void RTLGeneratorCalloutPass::runOnOperation() {
+  ModuleOp root = getOperation();
+  std::string genExecutableName, genExecutablePath;
+  std::vector<std::string> genOptions;
+  parseProgramArgs(genExecutable, genExecArgs, genExecutableName,
+                   genExecutablePath, genOptions);
+
+  OpBuilder builder(root->getContext());
+  auto genMemExe =
```
Comment (darthscsi) [inline lib/Dialect/SV/Transforms/GeneratorCallout.cpp:75]: Deal with the executable in the executable finding function you split off from the parse function.

File: lib/Dialect/SV/Transforms/GeneratorCallout.cpp [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+      llvm::sys::findProgramByName(genExecutableName, {genExecutablePath});
+  // If the executable was not found in the provided path (when the path is
+  // empty), then search it in the $PATH.
+  if (!genMemExe)
+    genMemExe = llvm::sys::findProgramByName(genExecutableName);
+  // If cannot find the executable, then nothing to do return.
+  if (!genMemExe)
+    return;
+  SmallVector<Operation *> opsToRemove;
+  for (auto &op : *root.getBody()) {
```
Comment (darthscsi) [inline lib/Dialect/SV/Transforms/GeneratorCallout.cpp:85]: there is an llvm:preinc iterator adaptor which will let you directly erase the op you are looking at in the loop, without needing a side vector.

File: lib/Dialect/SV/Transforms/GeneratorCallout.cpp [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+    if (auto modGenOp = dyn_cast<RTLModuleGeneratedOp>(op)) {
+      // Get the corresponding schema associated with this generated op.
+      if (auto genSchema =
+              dyn_cast<RTLGeneratorSchemaOp>(modGenOp.getGeneratorKindOp())) {
+        if (genSchema.descriptor().str() != schemaName)
+          continue;
+        std::vector<std::string> generatorArgs;
```
Comment (darthscsi) [inline lib/Dialect/SV/Transforms/GeneratorCallout.cpp:92]: SmallVector

File: lib/Dialect/SV/Transforms/GeneratorCallout.cpp [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+          auto fileContent = (*bufferRead)->getBuffer().str();
+          builder.setInsertionPointAfter(&op);
+          RTLModuleExternOp extMod = builder.create<rtl::RTLModuleExternOp>(
+              modGenOp.getLoc(), modGenOp.getVerilogModuleNameAttr(),
+              modGenOp.getPorts());
+          // Attach an attribute to which file the definition of the external
+          // module exists in.
+          extMod->setAttr("filenames", builder.getStringAttr(fileContent));
+          opsToRemove.push_back(modGenOp);
+          // builder.create<sv::VerbatimOp>(op.getLoc(), "`include " +
```
Comment (darthscsi) [inline lib/Dialect/SV/Transforms/GeneratorCallout.cpp:147]: remove temporary code before commit.

File: lib/Dialect/SV/Transforms/GeneratorCallout.cpp [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+          // Get the value for the corresponding port name.
+          // Assumption: the value always exists.
+          auto v = modGenOp->getAttr(portName);
+          if (auto intV = v.dyn_cast<IntegerAttr>())
+            generatorArgs.push_back(intV.getValue().toString(10, false));
+          else if (auto strV = v.dyn_cast<StringAttr>())
+            generatorArgs.push_back(strV.getValue().str());
+        }
+        generatorArgs.push_back(" ");
+        std::vector<StringRef> generatorArgStrRef;
```
Comment (darthscsi) [inline lib/Dialect/SV/Transforms/GeneratorCallout.cpp:120]: SmallVector.  Also, you should be able to construct this earlier.

File: test/Dialect/SV/rtl-generator-callout.mlir [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+  rtl.module.generated @sampleModuleName, @SchemaVar(%rd_clock_0: i1, %rd_en_0: i1, %rd_addr_0: i4,
+  %rd_clock_1: i1, %rd_en_1: i1, %rd_addr_1: i4) -> (%rd_data_0: i16, %rd_data_1: i16) attributes {port1 = 10 : i64, port2 = 2 : i32}
+
+  // CHECK:  rtl.module.extern @sampleModuleName(%rd_clock_0: i1, %rd_en_0: i1, %rd_addr_0: i4, %rd_clock_1: i1, %rd_en_1: i1, %rd_addr_1: i4) -> (%rd_data_0: i16, %rd_data_1: i16) attributes {filenames =  "file1.v,file2.v,file3.v,file4.v --moduleName sampleModuleName --port1 10 --port2 2  "}
+
+  rtl.generator.schema @SchemaVar1, "Schema_Name1", ["port1", "port2"]
+  rtl.module.generated @sampleModuleName1, @SchemaVar1(%rd_clock_0: i1, %rd_en_0: i1, %rd_addr_0: i4, %rd_clock_1: i1, %rd_en_1: i1, %rd_addr_1: i4) -> (%rd_data_0: i16, %rd_data_1: i16) attributes {port1 = 10 : i64, port2 = 2 : i32}
+  // CHECK: rtl.module.generated @sampleModuleName1, @SchemaVar1(%rd_clock_0: i1, %rd_en_0: i1, %rd_addr_0: i4, %rd_clock_1: i1, %rd_en_1: i1, %rd_addr_1: i4) -> (%rd_data_0: i16, %rd_data_1: i16) attributes {port1 = 10 : i64, port2 = 2 : i32}
+  sv.verbatim "// Standard header to adapt well known macros to our needs."
```
Comment (darthscsi) [inline test/Dialect/SV/rtl-generator-callout.mlir:13]: You don't need the verbatim stuff to test this.

File: test/Dialect/SV/rtl-generator-callout.mlir [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+  sv.verbatim ""
+  rtl.module @top_mod(%clock: i1, %reset: i1, %r0en: i1) -> (%data: i16) {
```
Comment (darthscsi) [inline test/Dialect/SV/rtl-generator-callout.mlir:15]: you probalby don't even need the module to test it.

File: lib/Dialect/SV/Transforms/GeneratorCallout.cpp (hunk 10 above)
Comment (prithayan) [reply lib/Dialect/SV/Transforms/GeneratorCallout.cpp]: Still need a separate loop to create `StringRef` vector, can't use the temporary string inside the loop.

**Comment by prithayan** [comment]
Addressed the review comments, 

- Moved the `findProgramByname` to parsing function
- Use `StringRef::split` for parsing the program args
- Use `make_early_inc_range` iterators to remove ops in the same loop
- Update `std::vector` to `SmallVector`

**Comment by prithayan** [comment]
Addressed all the requested changes, 
Added a todo, for the only pending item is to find an LLVM API for parsing the path out of the executable in a platform independent manner, 
`auto pathLoc = genExec.find_last_of("/\\");`

**Comment by lattner** [comment]
Hi @prithayan,

I finally got a chance to take a look at this, and made some minor changes in 286dc8d3 to just move the code around a bit.  That said, I have a few requests for you:

- We need to resolve the TODO's in the file, importantly the error cases.  I think that each of these (if the exec fails, if the memory buffer can't be read, the generator isn't a `RTLGeneratorSchemaOp`, when the fields of `requiredAttrs()` have unexpected type) should cause an emitError on the generator op instead of being silently forgotten about.
 - The TODO's in parseProgramArgs can be resolved by using the routines in llvm/Support/Path.h
 - Why does the design of this rely on a single global program set by the `execName` command line option?  I thought the design for this was to put this information into the `RTLGeneratorSchemaOp` node.  That would allow having multiple different generators without a multiplexer script.
 - Similarly why is the output filename specified on the command line?  I think we should expect that all generators take a `-o` command with a path, and that *we* generate a path to use (using `getPotentiallyUniqueTempFileName` in llvm/Support/FileSystem.h).  This will allow us to run generators in parallel safely someday.

Lemme know if you'd like me to file an issue track this instead,

-Chris

**Comment by lattner** [comment]
Thx Prithayan!

**Comment by prithayan** [comment]
Created a PR to address the issues, https://github.com/llvm/circt/pull/1038
1. Handled all the error cases properly. All the other assumptions are ensured by the IR.
2. Used the temporary file creation for standard out. Not going with the ``-o`` option, since  we are using just the ``echo`` in our lit tests as a dummy generator program. 
3. Also, we need the global ``execName`` command line option since we might need multiple memory generators for a single schema, we can't register a single executable with the schema.

**Comment by lattner** [comment]
Why not use multiple schema's for different generators?

**Pull request merged**